man1/gitodo.1
.TH GITODO 1 "2017-11-11" "Git-TODO" "A personal task manager"
.SH NAME
gitodo \- a personal task manager
.SH SYNOPSIS
\fBgitodo\fP
.br
\fBgitodo\fP \fB\-\-list\fP \fIid\fP [\fIid ...\fP]
.br
\fBgitodo\fP \fB\-\-edit\fP \fIid\fP [\fIid ...\fP]
.br
\fBgitodo\fP \fB\-\-print\fP \fIid\fP [\fIid ...\fP]
.br
\fBgitodo\fP \fB\-\-delete\fP \fIid\fP [\fIid ...\fP]
.br
\fBgitodo\fP \fB\-\-new\fP [\fIn\fP]
.br
\fBgitodo\fP \fB\-\-new\-from\-files\fP \fIfile\fP [\fIfile\fP ...]
.br
\fBgitodo\fP \fB\-\-body\fP \fIterm\fP
.br
\fBgitodo\fP \fB\-\-cron\fP
.br
\fBgitodo\fP \fB\-\-cron\-outdated\fP
.br
\fBgitodo\fP \fB\-\-count\fP
.br
\fBgitodo\fP \fB\-\-raw\fP
.br
\fBgitodo\fP \fB\-\-rotting\fP [\fIn\fP]
.SH DESCRIPTION
gitodo is a simple personal task manager for your command line.
.P
You can do the following:
.P
.IP \(bu 4
Add, edit and list your tasks. Essentially, that's a TODO list.
.IP \(bu 4
Prioritize some tasks. Negative priorities mean a higher urgency.
.IP \(bu 4
Set a deadline for tasks.
.IP \(bu 4
Be able to synchronize your tasks with other machines because they're
kept in a Git repository.
.P
When listing tasks, they're first sorted by importance then by deadline.
Typical output from the script might look like this:
.P
\f(CW
.nf
 S   Pri      Deadline         ID    Subject
---+-----+------------------+------+-----------------------------------
   | -15 |                  |    1 | Mail to Jane: Dinner
   |  -3 |                  |    2 | Christmas presents
 D |   0 |            22:00 |    9 | Remove garbage from your bed
 C |   0 |            23:30 |    8 | Go to sleep -- yes, it's important
   |   0 | 2010-12-10 23:59 |   10 | Still awake?
   |   0 | 2010-12-13       |    6 | Prepare talk about Git
   |   0 | 2010-12-24       |    7 | Remove SVN from all computers
   |   0 |                  |    4 | Wash dishes
   |   0 |                  |    5 | Buy new milk
   |   5 |                  |    3 | Wash the car
.fi
\fP
.P
Note the order: Items with a higher importance come first. If two items
have the same importance, they're sorted by their deadline. Deadlines
can be a date, a time (meaning \(lqtoday\(rq) or both. If there's no
deadline, the task is considered to be \(lqnot important\(rq and is
moved to the bottom.
.P
If the deadline for a task has already passed, it's marked with a
\fBD\fP at the beginning of the line. If the deadline is close (usually
less than 1 hour left, can be set), the task is marked with a \fBC\fP.
.SH OPTIONS
\fBgitodo\fP is meant to be used on the command line. To make things
easier for you, you can leave out leading dashes. That is, instead of
using \fB\-\-list\fP you could just write \fBlist\fP or even \fBl\fP
because \fB\-l\fP is the short alias for \fB\-\-list\fP.
.P
As you'll see, all important commands have a short alias.
.TP
\fB\-l\fP, \fB\-\-list\fP \fIid [id ...]\fP
List all matching items.
.TP
\fB\-e\fP, \fB\-\-edit\fP \fIid [id ...]\fP
Edit all matching items using $\fBEDITOR\fP, one after another.
.TP
\fB\-p\fP, \fB\-\-print\fP \fIid [id ...]\fP
Print all matching items including their body to \fBstdout\fP.
.TP
\fB\-n\fP, \fB\-\-new\fP [\fIn\fP]
Add \fIn\fP new tasks. You'll be put in your $\fBEDITOR\fP to do so. If
you don't specify \fIn\fP, exactly one new task will be created.
.TP
\fB\-f\fP, \fB\-\-new\-from\-files\fP \fIfile\fP [\fIfile\fP ...]
Creates a new task from each file. You can use this in scripts. Note
that the editor will be opened for each new task.
.TP
\fB\-d\fP, \fB\-\-delete\fP \fIid [id ...]\fP
Delete all matching items.
.TP
\fB\-b\fP, \fB\-\-body\fP \fIterm\fP
Search in bodies of items and print matching items (including their
bodies). \fIterm\fP must be a regular expressions understood by Ruby (no
modifiers allowed). See \fBSEARCHING\fP.
.TP
\fB\-c\fP, \fB\-\-cron\fP
Show a list of outdated tasks and tasks close to the deadline, suitable
for a cron job.
.TP
\fB\-o\fP, \fB\-\-cron\-outdated\fP
Show a list of outdated tasks, suitable for a cron job.
.TP
\fB\-\-count\fP
Count tasks: Dead tasks, tasks close to deadline, tasks without a
deadline, and total number of tasks.
.TP
\fB\-\-raw\fP
Don't do pretty printing. This option is meant for scripting.
.TP
\fB\-\-rotting\fP [\fIn\fP]
When using \fBgitodo\fP for a long time, you might lose track of some
items \(em especially those without a deadline. \fB\-\-rotting\fP lists
all tasks which have not been changed for \fIn\fP months. Default is 3
months. This only affects tasks without a deadline.
.P
\fIid\fP can be an explicit ID or string beginning with \fB:/\fP (see
\fBSEARCHING\fP) or a special identifier: \fBdead\fP means all outdated
items, \fBclose\fP means all items close to their deadline, \fBtoday\fP
and \fBweek\fP means all items whose deadline is today or in the current
week.
.SH SYNTAX
Items are simple text files. They look like this:
.P
\f(CW
.nf
    what: Subject of the task
    when: 2010-12-10 08:30
    prio: 2
    warn: 4

    Some other lines describing the task in detail.
.fi
\fP
.P
Deadline and priority are optional. In fact, only a subject line is
mandatory. The default priority is 0 and no deadline means 2037-01-01
00:00:00. ;-) A negative number means higher priority (just like Linux'
\fBnice\fP(1) values). The priority is meant to range from -99 to 99.
.P
\fBwhen\fP can be anything that is understood by Ruby's
\fBTime.parse()\fP function. It can also denote a relative point in
time, for example \fB+1h\fP. The following suffixes are understood:
.P
.TP
\fBh\fP
Hours.
.TP
\fBd\fP
Days. This is equal to adding 24 hours.
.TP
\fBw\fP
Weeks. This is equal to adding 7 days.
.TP
\fBm\fP
Months. This is equal to adding 30 days.
.P
There must not be any whitespace, so \fB+3m\fP is valid but \fB+ 1h\fP
is not.
.P
\fBwarn: 4\fP means that this task gets marked as \(lqclose to
deadline\(rq four hours before the deadline will be reached. If this
fields is left out, it defaults to one hour.
.P
\fBsubject:\fP is an alias for \fBwhat:\fP and \fBdead:\fP is an alias
for \fBwhen:\fP.
.P
If there's a line containing only \fBnocron\fP, then this item is not
shown when \fBgitodo\fP is called with \fB\-\-cron\fP or
\fB\-\-cron-outdated\fP.
.P
To help you procrastinate, \fBgitodo\fP knows a special item type: Items
with a priority less than 99 (i.e., 100, 101, 102, ...) which do have a
deadline set are hidden from listings. Such items are not included in
\fB\-\-cron\fP output and \fB\-\-count\fP, either. They're only shown if
they are close to their deadline or even outdated.
.P
This is meant for TODO items that are so unimportant that you don't want
to be bothered by them \(em until it really is about time.
.P
\f(CW
.nf
    what: Wash the car
    dead: 2012-06-15 12:00
    prio: 100

    I really should do this. Some day.
.fi
\fP
.P
If you do want to see such items, you can set the environment variable
$\fBGITODO_SHOW_UNIMPORTANT\fP.
.SH "TASK ITEM FILES, THE REPOSITORY AND SYNCHING"
All your task items are kept in a single directory. Every change of your
task list will be automatically committed to that repository. By
default, \fI$XDG_DATA_HOME/gitodo.items\fP will be used or
\fI$HOME/.local/share/gitodo.items\fP if $\fBXDG_DATA_HOME\fP is not
set. Use the environment variable $\fBGITODO_DATA\fP to change this. If
it's set, then this directory is used.
.P
\fBImportant note:\fP Your items directory must exist and it must
already be initialized as a Git repository. \fBgitodo\fP will not do
this for you.
.P
Common task items will be named like \fIi0123\fP: That would be the file
for the 123rd item. Note that new items will get a random free ID
ranging from 0 to 9999. This is done to avoid merge conflicts.
.P
Add the items repo to your synching mechanism if you want them to be
synched.
.SH "HANDLING EMPTY ITEMS"
\fBgitodo\fP will automatically remove empty items from your repository.
\(lqEmpty\(rq means a completely empty file or a file containing only
whitespace. This is a useful housekeeping mechanism.
.P
It's also relevant if you create new items using \fBgitodo \-\-new\fP.
If you change your mind (\(lqnah, actually, I don't care about this TODO
item\(rq), then just delete all lines while you're still in the editor
and the new item will be dropped. This is similar to commit messages in
\fBgit\fP: When entering an empty commit message, no commit will be
made.
.SH SEARCHING
Instead of giving explicit IDs, you can also use Ruby regular
expressions to search the subjects of your items. The following command
edits all items matching \(lqjane\(rq or \(lqdiane\(rq, your TODO item
with ID 432 and all items matching \(lqgnu\(rq:
.P
\f(CW
.nf
    $ gitodo -e ':/(j|di)ane' 432 ':/gnu'
.fi
\fP
.P
Searching using \fB:/\fP as a prefix works with the \fB\-\-edit\fP,
\fB\-\-print\fP, \fB\-\-delete\fP and \fB\-\-list\fP commands. Let me
repeat: This only searches in \fBsubject lines\fB.
.P
To filter your list of TODO items, you can use the \fB\-\-list\fP
command. Again, you can either start your search term with \fB:/\fP to
perform a searching using Ruby regular expressions:
.P
\f(CW
.nf
    $ gitodo --list :/time
     S   Pri      Deadline         ID    Subject
    ---+-----+------------------+------+-----------------
       |   0 | 2012-07-13 11:01 | 1156 | time_t party
.fi
\fP
.P
Or, you can supply one or more ids and even mix both variants:
.P
\f(CW
.nf
    $ gitodo --list 7458 2436 :/time
     S   Pri      Deadline         ID    Subject
    ---+-----+------------------+------+-----------------
       |   0 | 2012-07-01       | 7458 | lico-update
       |   0 | 2012-07-13 11:01 | 1156 | time_t party
       |   1 |                  | 2436 | aoi patches
.fi
\fP
.P
To search in \fBbodies\fP of items, you can use \fB\-\-body\fP. Note:
Don't prefix your search term with \fB:/\fP when using \fB\-\-body\fP.
The \fB:/\fP is only needed to distinguish IDs from search terms. As
\fB\-\-body\fP does not understand IDs, there's no need for \fB:/\fP.
.P
\f(CW
.nf
    $ gitodo --body party
     1156
    ------
    what: time_t party
    when: 2012-07-13 11:01
    prio: 50

    At 11:01:20 UTC on July 13, 2012, the Unix time number will
    reach 0x50000000 (1,342,177,280 seconds). With the exception of
    locations in time zones UTC−12:00, UTC+13:00, and UTC+14:00,
    this also happens on a Friday the 13th.

    http://en.wikipedia.org/wiki/Unix_time
.fi
\fP
.P
Again, you can use Ruby regexes with \fB\-\-body\fP.
.P
Searching is always case insensitive. For example, if you specify
\fI:/foo\fP, internally \fI/foo/i\fP will be used.
.SH "USING GITODO WITH HIGHCAL"
The \fBhighcal\fP(1) script is an additional and optional Ruby script.
It's purpose is to create a calendar similar to \fBcal\fP(1). The script
also reads from \fBstdin\fP a list of dates which will be highlighted.
On Unix-like systems, it can be easily combined with gitodo:
.P
\f(CW
.nf
    $ gitodo --raw | cut -d' ' -f2 | highcal
.fi
\fP
.P
You'll see the current month plus the two following months. The current
day is marked as well as all deadlines for TODO items (as long as there
is one).
.SH PORTABILITY
\fBgitodo\fP once was a shell script that could be run in the
\fBmsysgit\fP (http://code.google.com/p/msysgit/) environment on
Windows. This is no longer true. \fBgitodo\fP is now implemented in Ruby
1.9 and only tested on current versions of Arch Linux.
.P
Anyway, a little advice: If you ever think about putting the task items
repo on a USB stick and if you want to use that repo from both Windows
and GNU, make sure to \fBdisable\fP msysgit's \(lqautocrlf\(rq feature.
For more information, see:
.P
http://www.tigraine.at/2010/02/03/disable-autcrlf-in-msysgit/
.P
If you leave it enabled, it's going to be nasty.
.SH ENVIRONMENT
.TP
\fBGITODO_EDITOR\fP, \fBEDITOR\fP
\fBGITODO_EDITOR\fP has priority over \fBEDITOR\fP. They point to the
editor you wish to use to edit your items. Defaults to \fIvim\fP if both
are unset.
.TP
.B GITODO_EDIT_SINGLE
If this variable is set, the editor will only be called once for all
edit operations. For example, suppose you do a \fIgitodo e :/foo\fP and
\fI:/foo\fP matches the items \fIi1000\fP and \fIi2000\fP. If
\fBGITODO_EDIT_SINGLE\fP is set, only one instance of your editor will
be started, e.g. \fIvim i1000 i2000\fP. If the variable is \fBNOT\fP
set, then several instances will be started, e.g. \fIvim i1000\fP and
then \fIvim i2000\fP after the first Vim has finished. The use case of
\fBGITODO_EDIT_SINGLE\fP is to make use of Vim buffers or using \fIvim
-p\fP as an editor.
.TP
.B GITODO_DATA
If this variable is set, it must point to a directory where your git
repository will be stored. You can use this to override the default
location (see \fBFILES\fP below).
.TP
.B GITODO_FORCE_COLOR
If this variable is set, color is used in any case, even if stdout is
not a terminal.
.TP
.B GITODO_SHOW_UNIMPORTANT
If this variable is set, filtering of \(lqunimportant items\(rq (see
above) will be disabled.
.SH FILES
.TP
.PD 0
\fI$XDG_DATA_HOME/gitodo.items/\fP
.TP
.PD
\fI$HOME/.local/share/gitodo.items/\fP
Default per-user data directory if $\fBGITODO_DATA\fP is not set. The
$\fBHOME\fP variant is only used if $\fBXDG_DATA_HOME\fP is not set.
.SH BUGS
Currently, no bugs are known. If you find one, we invite you to report
it at the GitHub Issue tracker (http://github.com/vain/gitodo/issues).
.SH LICENSE
\fBgitodo\fP is released under the MIT license. See the accompanying
\fILICENSE\fP file.
.SH HISTORY
\fBgitodo\fP was originally written by Peter Hofmann. The project was
started in December 2010.
.SH "SEE ALSO"
.BR highcal (1),
.BR git (1),
.BR ruby (1).


CHANGES
Release history for gitodo

next
  [Changed]
  - Header fields will now be normalized. "dead" will automatically be
    rewritten as "when" and "subject" will be changed to "what". Please
    note that these aliases are now considered deprecated and might be
    removed in the following version.

  [Added]
  - The "when" field can now include relative value, e.g. "+1h".

v17.11  2017-11-16
  [Changed]
  - "--count" now also prints the number of tasks that don't have a
    deadline set.

v17.09  2017-09-16
  [LICENSING]
  - The project has been relicensed under the MIT license (more
    precisely called "Expat License", but everyone calls it "MIT
    license" these days).

v17.04  2017-04-17
  [Fixed]
  - We now use Time.parse() to parse the "when" field and any input to
    highcal. Unlike DateTime.parse(), which we used before, it returns
    the proper time zone. (Plus, we don't need a weird hackaround for
    time zones anymore.)

v16.09  2016-09-03
  [Changed]
  - Use "git rev-parse HEAD" to determine if GITODO_DATA is a git
    repository instead of simply appending ".git" to the path. This
    allows GITODO_DATA to be a subdirectory of a git repo -- which means
    that you can use one single repo and have multiple "categories" of
    items, organized in subdirectories.

v15.09 and earlier releases
  - Sorry, the CHANGES file did not exist back then. Please refer to the
    Git log.


gitodo.rb
#!/usr/bin/env ruby
# encoding: UTF-8

require 'date'

module Colors
	@table = {
		:dead      => ['tput bold; tput setaf 1', nil],
		:close     => ['tput setaf 1', nil],
		:very_high => ['tput bold; tput setaf 1', nil],
		:high      => ['tput setaf 1', nil],
		:low       => ['tput setaf 4', nil],
		:very_low  => ['tput setaf 2', nil],
		:header    => ['tput bold', nil],
		:reset     => ['tput sgr0', nil],
	}

	def Colors.[](which)
		if !STDOUT.tty? && !ENV['GITODO_FORCE_COLOR']
			return ''
		end

		# Lazy allocation to save useless forks.
		if @table[which][1] == nil
			@table[which][1] = `#{@table[which][0]}`
		end

		return @table[which][1]
	end
end

class Item
	include Comparable

	attr_accessor :id
	attr_accessor :filename
	attr_accessor :prio
	attr_accessor :when
	attr_accessor :warn
	attr_accessor :what
	attr_accessor :nocron
	attr_accessor :content
	attr_accessor :timetype

	def initialize
		@id = 0
		@filename = ''
		@prio = 0
		@when = Time.local(2037)
		@warn = 1
		@what = ''
		@nocron = false
		@content = ''
		@timetype = :none
	end

	def id_fmt
		'%04d' % @id
	end

	def classify
		now = Time.now
		if now > @when
			return :dead
		elsif now > @when - @warn * 3600
			return :close
		else
			return :normal
		end
	end

	def timetype_num
		case @timetype
		when :time
			0
		when :date
			1
		when :datetime
			2
		else
			3
		end
	end

	def show_brief
		case classify
		when :dead
			print(Colors[:dead])
			print('! ')
		when :close
			print(Colors[:close])
			print('* ')
		else
			print('  ')
		end

		print(Colors[:reset])

		if prio < -5
			print(Colors[:very_high])
		elsif prio < 0
			print(Colors[:high])
		elsif prio > 5
			print(Colors[:low])
		elsif prio > 0
			print(Colors[:very_low])
		end

		printf('[%3s]', @prio.to_s)

		print(' [')
		case @timetype
		when :none
			print('                   ')
		when :time
			print('           ' + @when.strftime('%T'))
		when :date
			print(@when.strftime('%F') + '         ')
		when :datetime
			print(@when.strftime('%F %T'))
		end
		print(']')
		printf(' [%4s]', @id.to_s)
		print(' ' + @what)
		print(Colors[:reset])

		puts
	end

	def show_detail
		print(Colors[:header])
		puts(id_fmt)
		puts(id_fmt.gsub(/./, '-'))
		print(Colors[:reset])
		puts

		puts(@content)

		puts
	end

	def show_raw
		print("#{@prio} #{@when.strftime('%F %T %s')} ")
		print("#{@warn} #{@nocron ? '1' : '0'} #{timetype_num} ")
		print("#{@id} #{@what}")

		puts
	end

	def <=>(other)
		[@prio, @when] <=> [other.prio, other.when]
	end
end

def load_all_items
	items = []

	Dir.glob('i*').each do |filename|
		f = File.new(filename, 'r')
		lines = f.readlines()
		f.close()

		item = Item.new

		filename =~ /.*i0*([[:digit:]]+)$/
		item.id = $1
		item.filename = filename

		item.content = lines.join('')

		lines.each do |line|
			if line =~ /^(prio): (.*)/
				item.prio = $2.to_i
			elsif line =~ /^(when|dead): (.*)/
				# XXX: Hackaround to be able to use DateTime.parse in
				# the local timezone.
				item.when = DateTime.parse($2)
				item.when -= Time.now.utc_offset / 86400.0
				item.when = item.when.to_time

				spaces = $2.count(' ')
				colons = $2.count(':')
				case spaces
				when 0
					if colons > 0
						item.timetype = :time
					else
						item.timetype = :date
					end
				else
					item.timetype = :datetime
				end
			elsif line =~ /^(warn): (.*)/
				item.warn = $2.to_i
			elsif line =~ /^(what|subject): (.*)/
				item.what = $2
			elsif line == 'nocron'
				item.nocron = true
			end
		end

		if item.prio <= 99 || ENV['GITODO_SHOW_UNIMPORTANT']
			items << item
		end
	end

	items.sort!

	return items
end

def check_and_chdir_repo(dir)
	if Dir.exists?(dir + '/.git')
		Dir.chdir(dir)
	end
end

def print_header(indent=false)
	print(Colors[:header])
	puts((indent ? "\t" : '') + 'O Prio        Deadline          ID   Subject')
	puts((indent ? "\t" : '') + '- ----- --------------------- ------ -------')
	print(Colors[:reset])
end

def list_items(items, indent=false)
	print_header(indent)
	items.each do |i|
		if indent
			print("\t")
		end
		i.show_brief
	end
end

def print_items(items)
	items.each do |i|
		i.show_detail
	end
end

def filter_ids(items, filter)
	# TODO: This works but feels very odd.
	items.select do |i|
		match = false
		filter.each do |f|
			if f.start_with?(':/')
				if i.what =~ /#{f[2..-1]}/i
					match = true
					next
				end
			elsif f == 'dead'
				if i.classify == :dead
					match = true
					next
				end
			elsif f == 'close'
				if i.classify == :close
					match = true
					next
				end
			else
				if i.id == f
					match = true
					next
				end
			end
		end
		match
	end
end

def filter_bodies(items, filter)
	items.select { |i| i.content =~ /#{filter}/i }
end

def filter_rotting(items, months)
	now = Time.now
	items.select do |i|
		lchange = `git log '--pretty=format:%ct' -n 1 -- "#{i.filename}"`
		lchange = Time.at(lchange.to_i)
		now - lchange > months * 60 * 60 * 24 * 30
	end
end

def print_cron(items, selection)
	puts('OUTDATED TASKS:')
	puts('===============')
	puts
	list_items(filter_ids(items, ['dead']), true)

	if selection == :all
		puts
		puts

		puts('DEADLINE CLOSE:')
		puts('===============')
		puts
		list_items(filter_ids(items, ['close']), true)
	end
end

def new_items(editor, how_many)
	how_many.times do
		filename = 'i' + ('%04d' % rand(0..9999))
		while File.exists? filename
			filename = 'i' + ('%04d' % rand(0..9999))
		end

		system "#{editor} #{filename}"
	end
end

def delete_items(items)
	items.each do |i|
		File.delete(i.filename)
	end
end

def commit
	system 'git add .'
	system 'git commit -a -m "Auto-commit $(hostname)."'
end


default_prefix = ENV['XDG_DATA_HOME'] || (ENV['HOME'] + '/.local/share')
datadir = ENV['GITODO_DATA'] || (default_prefix + '/gitodo.items')
editor = ENV['EDITOR'] || 'vim'

if !check_and_chdir_repo(datadir)
	$stderr.puts "`#{datadir}' is not a Git repository."
	exit 1
end
items = load_all_items

if ARGV.size == 0
	list_items(items)
	exit
end

opt = ARGV[0].sub(/^-+/, '')
ARGV.shift
case opt
when 'new', 'n'
	new_items(editor, ARGV[0] ? ARGV[0].to_i : 1)
	commit
when 'list', 'l'
	list_items(filter_ids(items, ARGV))
when 'print', 'p'
	print_items(filter_ids(items, ARGV))
when 'body', 'b'
	print_items(filter_bodies(items, ARGV[0]))
when 'edit', 'e'
	filter_ids(items, ARGV).each do |i|
		system "#{editor} #{i.filename}"
	end
	commit
when 'delete', 'd'
	filter_ids(items, ARGV).each do |i|
		puts "Deleting #{i.id_fmt}: #{i.what}"
		File.delete(i.filename)
	end
	commit
when 'count'
	total = items.size
	dead = items.count { |i| i.classify == :dead }
	close = items.count { |i| i.classify == :close }
	puts("#{dead} #{close} #{total}")
when 'cron', 'c'
	print_cron(items, :all)
when 'cron-outdated', 'o'
	print_cron(items, :no_close)
when 'raw'
	items.each { |i| i.show_raw }
when 'rotting'
	list_items(filter_rotting(items, ARGV[0] ? ARGV[0].to_i : 3))
else
	puts 'Invalid option. See manpage.'
end
